### PR TITLE
implement the `L25` intermediate language

### PR DESCRIPTION
**Pull request by zerbina**
## Summary

Add the `L25` language and integrate it into the pass pipeline. The
`L25` features a flat procedure structure and goto-esque control-flow,
without using a basic block structure and SSA yet. It comes after `L30`
and before `L4`.

## Details

The new IL is planned to be the first in a series of ILs that all use
flat procedure bodies with goto-esque control-flow constructs. This
structure works well for the stage during compilation where data- and
control-flow analysis is needed, but where the live range(s) of locals
can still change.

Future passes that are currently planned to use this structure are:
borrow checking, cursor inference, move analysis, destructor injection,
and inlining.

### Implementation

`pass30` is effectively split into two passes. Turning the structure
control-flow constructs (i.e., `If`, `Case`, `Loop`, etc.) stays in
`pass30`, while the data-flow analysis and SSA transformation moves to
`pass25` (without being modified).

Most of the `pass25` are former `pass30` tests (those concerning data-
flow analysis and SSA transformation), the rest are new tests covering
the basic `L25` to `L4` translation.

---

## To-Do
* [x] add more tests. All IL features should be covered
* [x] implement `Except` support in `pass7`

## Notes for Reviewers
* this PR is part of the pass rework project

### CODE REVIEW

**Comment by zerbina** [comment]
I'm going to wait with merging this addition until the listed passes are actually needed. While I think the language is a solid choice for the state purpose, it's possible that there's a better approach that I haven't yet considered, so I'm not rushing to add the IL (just to remove it again later, should a better alternative emerge).

The first pass out of the listed one that I think will be needed is drop injection.

**Comment by zerbina** [comment]
For making sure the pass is working correctly, and to also get some feedback on performance, I've run `pass7` on `L7` code translated from the fully processed MIR produced by the NimSkull compiler for the `repl.nim` program (~2.5MB of packed nodes).

Besides discovering multiple issues with the data-flow analysis (which are fixed by #54), this also showed that the pass takes up too much time. In a normal debug build, the pass takes 30 seconds (!) to process all procedures, while in `-d:release` mode, it takes 4 seconds. Considering that the `repl.nim` is only a small to medium sized program (~1000 procedures), this is far too much time.

### Reasons for Slowness

In order of significance:

1. **Issue with `Changeset.replace`.** The changeset's node sequence is, for some reason, not moved into the builder, resulting in a full copy of the whole sequence

2. **`PackedSet` is slow.** Or at least the operations relevant to the data-flow analysis (i.e., union and intersection) are. In addition, a `PackedSet` instance itself has a very large static size (320 byte!), ballooning up the static size of `BBlock` to 688 byte! This quickly adds up, especially since there are usually a lot of basic blocks in a single procedure.

3. **Type lookup is slow.** Looking up a type via its index requires skipping over all predecessor nodes in the tree, which takes longer the further the index is away from the start.

4. **`PackedSet.len` is slow.** Especially if only used to test whether the set is empty or not.

Number 1 is easy to fix, and 2 and 4 can be addressed by using a `Table`-based sparse set implementation. With the aforementioned three things fixed, the pass only takes ~550ms in release mode, which - while a lot better - is still too long.

Reducing the number of basic blocks (by combining them where possible), reducing the maximum number of variables live at the same time (by adding "end of storage duration" markers to the `L7`), some general optimization to the pass itself, as well as improving type lookup efficiency (possibly through using a skip list) should together be able to shrink the pass' run time to somewhere below 100ms, which seems acceptable.

**Comment by zerbina** [comment]
Okay, `Except` support is now implemented and the fixes from `main` are merged. Some tests are missing, but otherwise the bulk of the work should be done.

---

I did quite a bit of testing with real-world code bases, and I'm now fairly certain that the structure and idea with `L7` is the right choice. There are a few things that need to be changed (relative to the current state), namely:
* **"end of life" marker for locals**. Something like `StorageEnd` or `StorageDead`), in order to mark them as dead. Right now, locals have to be considered alive after their first write (or possible write), which leads to locals that have their address taken usually living much longer than necessary.

* **A `Move` operator.** It functions similar to `Copy`, with the addition of communicating that the source location is not used afterwards. An `L7` `Move` would be translated directly to an `L4` `Move`.

* **Requiring locals to be initialized prior to being used**. (Applies to both the `L10` and `L7`.) This would simplify some code, by removing the need for auto-spawning, and - more importantly - removes a case of undefined behaviour (i.e.: what's the content of an uninitialized local?). In case not initializing locals prior to their first use should be allowed in the source language - like it is in NimSkull -, there needs to be a separate pass that initializes the problematic locals with their type's default value.

This PR is only concerned with splitting up `pass10`, so the changes should happen via follow-up PRs.

File: passes/pass25.nim [hunk 1]
```diff
@@ -0,0 +1,522 @@
+## Lowers |L7| into the |L4| language. The pass needs to split the statements
```
Comment (saem) [inline passes/pass25.nim:1]:
```suggestion
## Lowers |L25| into the |L4| language. The pass needs to split the statements
```

**Review by saem — APPROVED** [review]
a few minor doc suggestions.

File: passes/builders.nim [hunk 2]
```diff
@@ -60,6 +72,23 @@ func copyFrom*[T](bu: var Builder[T], tree: PackedTree[T], n: NodeIndex) =
...
+  ## Inserts a placeholder node into the current tree. Using the returned
+  ## `Placeholder <#Placeholder>_` value, the placeholder node must later be
+  ## replaced with the correct node.
+  inc bu.numPlaceholders
+  if bu.parent != -1:
+    inc bu.buf[bu.parent].val
+  bu.buf.add TreeNode[T]() # add a placeholder node
+  Placeholder(pos: bu.buf.high)
+
+proc replace*[T](bu: var Builder[T], p: sink Placeholder, n: TreeNode[T]) =
```
[the remark this answers is not in the document]
Comment (saem) [reply passes/builders.nim]: this is outdated

File: passes/lang7.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+exit -= (Continue <cont_name> <value> (List <cont_arg>*))
+      | (Loop <cont_name> (List <cont_arg>*))
+      | (Raise <value> <err_goto>)
+      | (CheckedCall <proc> <value>* <goto> <err_goto>)
+      | (CheckedCall <type_id> <value>+ <goto> <err_goto>)
+
+exit += (Continue <cont_name> <value>)
+      | (Loop <cont_name>)
+      | (Return <value>?)
+      | (Raise <value> <err_goto>)
```
[the remark this answers is not in the document]
Comment (saem) [reply passes/lang7.md]: outdated

**Pull request merged**